Dalamud/Interface/ManagedFontAtlas/Internals/FontHandle.cs

```diff
@@ -0,0 +1,295 @@
+
+    private readonly InterfaceManager interfaceManager;
+    private readonly List<IDisposable> pushedFonts = new(8);
+
+    private IFontHandleManager? manager;
+    private long lastCumulativePresentCalls;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="FontHandle"/> class.
+    /// </summary>
+    /// <param name="manager">An instance of <see cref="IFontHandleManager"/>.</param>
+    protected FontHandle(IFontHandleManager manager)
+    {
+        this.interfaceManager = Service<InterfaceManager>.Get();
+        this.manager = manager;
+    }
+
+    /// <inheritdoc/>
+    public event IFontHandle.ImFontChangedDelegate? ImFontChanged;
+
+    /// <summary>
+    /// Event to be called on the first <see cref="IDisposable.Dispose"/> call.
+    /// </summary>
+    protected event Action? Disposed;
+
+    /// <inheritdoc/>
+    public Exception? LoadException => this.Manager.Substance?.GetBuildException(this);
+
+    /// <inheritdoc/>
+    public bool Available => (this.Manager.Substance?.GetFontPtr(this) ?? default).IsNotNullAndLoaded();
+
+    /// <summary>
+    /// Gets the associated <see cref="IFontHandleManager"/>.
+    /// </summary>
+    /// <exception cref="ObjectDisposedException">When the object has already been disposed.</exception>
+    protected IFontHandleManager Manager => this.manager ?? throw new ObjectDisposedException(this.GetType().Name);
+
+    /// <inheritdoc/>
+    public void Dispose()
+    {
+        if (this.manager is null)
+            return;
+
+        this.Dispose(true);
+        GC.SuppressFinalize(this);
+    }
+
+    /// <summary>
+    /// Invokes <see cref="IFontHandle.ImFontChanged"/>.
+    /// </summary>
+    /// <param name="font">The font, locked during the call of <see cref="ImFontChanged"/>.</param>
+    public void InvokeImFontChanged(ILockedImFont font)
+    {
+        try
+        {
+            this.ImFontChanged?.Invoke(this, font);
+        }
+        catch (Exception e)
+        {
+            Log.Error(e, $"{nameof(this.InvokeImFontChanged)}: error");
+        }
+    }
+
+    /// <summary>
+    /// Obtains an instance of <see cref="ImFontPtr"/> corresponding to this font handle,
+    /// to be released after rendering the current frame.
+    /// </summary>
+    /// <returns>The font pointer, or default if unavailble.</returns>
+    /// <remarks>
+    /// Behavior is undefined on access outside the main thread.
+    /// </remarks>
+    public ImFontPtr LockUntilPostFrame()
+    {
+        if (this.TryLock(out _) is not { } locked)
+            return default;
+
+        if (!ThreadSafety.IsMainThread && nextNonMainThreadFontAccessWarningCheck < Environment.TickCount64)
+        {
+            nextNonMainThreadFontAccessWarningCheck =
+                Environment.TickCount64 + NonMainThreadFontAccessWarningCheckInterval;
+            var stack = new StackTrace();
+            if (Service<PluginManager>.GetNullable()?.FindCallingPlugin(stack) is { } plugin)
```

Comment:
This is extremely slow in my experience, did you see it cause any lag when rendering? Not sure if anything does enough pushing and popping for it to matter

Reply:
Ah, I missed that, thanks for clarifying!